### PR TITLE
feat: Nonce Verification

### LINKED ISSUE #107: feat: Nonce Verification 🎱
# Feature Request

**Describe the Feature Request**

Upon each transaction, the nonce should verified and ensured to be followed sequentially for each account address.

**Describe Preferred Solution**

One can simply take inspiration from the blockifier [implementation](https://github.com/tdelabro/blockifier/blob/no_std-support/crates/blockifier/src/transaction/account_transaction.rs#LL115C5-L133C6).

### PR DESCRIPTION
**Pull request by chirag-bgh**
# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing
- [ ] Other (please describe):

# What is the current behavior?

Issue Number: Closes #107 

# Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by EvolveArt** [comment]
Hey @chirag-bgh Please mark it as **DRAFT** when it's not ready for review.

You also need to integrate it on the current `execute` function, propagate error and add tests.

**Comment by abdelhamidbakhta** [comment]
> > hey @chirag-bgh just want to verify that you are still working on this issue
> 
> yes yes will do this asap. sorry for the inactivity

no worries at all, thank you

File: crates/primitives/starknet/tests/transaction_test.rs [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+use starknet_api::api_core::{
+    ClassHash, ContractAddress as StarknetContractAddress, ContractAddress, Nonce, PatriciaKey,
+};
+use starknet_api::hash::{StarkFelt, StarkHash};
+use starknet_api::transaction::{Fee, InvokeTransactionV1, TransactionHash, TransactionSignature, TransactionVersion};
+use starknet_api::{patricia_key, stark_felt};
+
+#[test]
+
+fn test_verify_nonce() {
```
Comment (EvolveArt) [inline crates/primitives/starknet/tests/transaction_test.rs:22]: please move test to `pallets/starknet/tests.rs` and use the mocked state constructed in `fn new_test_ext` in `mock.rs`

File: crates/primitives/starknet/tests/transaction_test.rs [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+                H256::from_str("0x00e7def693d16806ca2a2f398d8de5951344663ba77f340ed7a958da731872fc").unwrap(), /* Selector */
+                H256::from_str("0x0000000000000000000000000000000000000000000000000000000000000001").unwrap(), /* Length */
+                H256::from_str("0x0000000000000000000000000000000000000000000000000000000000000019").unwrap(), // Value
+            ],
+            contract_address_bytes,
+            contract_address_bytes,
+        ),
+        None,
+    );
+    let _txtype = TxType::InvokeTx;
```
Comment (EvolveArt) [inline crates/primitives/starknet/tests/transaction_test.rs:54]: unused ?

File: crates/primitives/starknet/src/transaction/mod.rs [hunk 3]
```diff
@@ -304,6 +313,35 @@ impl Transaction {
...
+        }
+        let currrent_nonce = state.get_nonce_at(account_tx_context.sender_address)?;
+        if currrent_nonce != account_tx_context.nonce {
+            return Err(TransactionExecutionError::InvalidNonce {
+                expected_nonce: account_tx_context.nonce,
+                actual_nonce: currrent_nonce,
+            });
+        }
+
+        Ok(state.increment_nonce(account_tx_context.sender_address)?)
```
Comment (EvolveArt) [inline crates/primitives/starknet/src/transaction/mod.rs:343]: this is probably not needed as already done in `apply_state_diffs` in `pallets/starknet/lib.rs`

**Review by EvolveArt — CHANGES_REQUESTED** [review]
left a few comments

File: crates/primitives/starknet/tests/transaction_test.rs (hunk 1 above)
Comment (chirag-bgh) [reply crates/primitives/starknet/tests/transaction_test.rs:22]: I couldn't get this. How to retrieve the mock state from the test environment created in `fn new_test_ext` ?

Comment (EvolveArt) [reply crates/primitives/starknet/tests/transaction_test.rs:22]: I suggest you look at other tests in the `starknet` pallet tests, you can just copy paste one and start from there

File: crates/pallets/starknet/src/tests.rs [hunk 4]
```diff
@@ -501,3 +515,76 @@ fn given_hardcoded_contract_run_storage_read_and_write_it_works() {
...
+                    H256::from_str("0x0000000000000000000000000000000000000000000000000000000000000001").unwrap(), /* Length */
+                    H256::from_str("0x0000000000000000000000000000000000000000000000000000000000000019").unwrap(), // Value
+                ],
+                contract_address_bytes,
+                contract_address_bytes,
+            ),
+            None,
+        ); 
+        
+        // create a state
```
Comment (chirag-bgh) [inline crates/pallets/starknet/src/tests.rs:560]: @EvolveArt I still couldn't find how to fetch the state.

Comment (EvolveArt) [reply crates/pallets/starknet/src/tests.rs:560]: @chirag-bgh That should be your test signature 

```rust
#[test]
fn test_name() {
    new_test_ext().execute_with(|| {
    /// Your test content there
})};
```

**Comment by EvolveArt** [comment]
CI is failing @chirag-bgh please make sure all tests run correctly

**Review by EvolveArt — APPROVED** [review]
lgtm

**Comment by tdelabro** [comment]
@chirag-bgh can you rebase your branch on the latest state?
@EvolveArt can we merge it when it's done ?

**Pull request merged**